google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1beta2/StreamWriterV2Test.java

```diff
@@ -288,6 +304,65 @@ public void serverCloseWhileRequestsInflight() throws Exception {
     }
 
     writer.close();
-    ;
+  }
+
+  @Test
+  public void testZeroMaxInflightRequests() throws Exception {
+    StreamWriterV2 writer =
+        StreamWriterV2.newBuilder(TEST_STREAM, client).setMaxInflightRequests(0).build();
+    testBigQueryWrite.addResponse(createAppendResponse(0));
+    verifyAppendIsBlocked(writer);
+    writer.close();
+  }
+
+  @Test
+  public void testZeroMaxInflightBytes() throws Exception {
+    StreamWriterV2 writer =
+        StreamWriterV2.newBuilder(TEST_STREAM, client).setMaxInflightBytes(0).build();
+    testBigQueryWrite.addResponse(createAppendResponse(0));
+    verifyAppendIsBlocked(writer);
+    writer.close();
+  }
+
+  @Test
+  public void testOneMaxInflightRequests() throws Exception {
+    StreamWriterV2 writer =
+        StreamWriterV2.newBuilder(TEST_STREAM, client).setMaxInflightRequests(1).build();
+    // Server will sleep 1 second before every response.
+    testBigQueryWrite.setResponseSleep(Duration.ofSeconds(1));
+    testBigQueryWrite.addResponse(createAppendResponse(0));
+
+    long appendStartTimeMs = System.currentTimeMillis();
+    ApiFuture<AppendRowsResponse> appendFuture1 = sendTestMessage(writer, new String[] {"A"});
+    long appendElapsedMs = System.currentTimeMillis() - appendStartTimeMs;
+    assertTrue(appendElapsedMs >= 1000);
+    assertEquals(0, appendFuture1.get().getAppendResult().getOffset().getValue());
+    writer.close();
+  }
+
+  @Test
+  public void testAppendsWithTinyMaxInflightBytes() throws Exception {
+    StreamWriterV2 writer =
+        StreamWriterV2.newBuilder(TEST_STREAM, client).setMaxInflightBytes(1).build();
+    // Server will sleep 100ms before every response.
+    testBigQueryWrite.setResponseSleep(Duration.ofMillis(100));
+    long appendCount = 10;
+    for (int i = 0; i < appendCount; i++) {
+      testBigQueryWrite.addResponse(createAppendResponse(i));
+    }
+
+    List<ApiFuture<AppendRowsResponse>> futures = new ArrayList<>();
+    long appendStartTimeMs = System.currentTimeMillis();
+    for (int i = 0; i < appendCount; i++) {
+      futures.add(sendTestMessage(writer, new String[] {String.valueOf(i)}));
+    }
+    long appendElapsedMs = System.currentTimeMillis() - appendStartTimeMs;
+    assertTrue(appendElapsedMs >= 1000);
+
+    for (int i = 0; i < appendCount; i++) {
+      assertEquals(i, futures.get(i).get().getAppendResult().getOffset().getValue());
```

Comment:
You code is safe, but maybe just verify the append is in order as an extra? You can assign an offset to request and get it from the server.

Reply:
Done.